src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/InMemoryEventStore.java

```diff
@@ -1,36 +1,75 @@
 package com.snowplowanalytics.snowplow.tracker.emitter;
 
 import com.snowplowanalytics.snowplow.tracker.payload.TrackerPayload;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.util.ArrayList;
-import java.util.concurrent.BlockingQueue;
-import java.util.concurrent.LinkedBlockingQueue;
-import java.util.List;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicLong;
 
 public class InMemoryEventStore implements EventStore {
-    public final BlockingQueue<TrackerPayload> eventBuffer = new LinkedBlockingQueue<>();
+    private static final Logger LOGGER = LoggerFactory.getLogger(InMemoryEventStore.class);
+    private final AtomicLong batchId = new AtomicLong(1);
+
+    public final LinkedBlockingDeque<TrackerPayload> eventBuffer;
+    public final ConcurrentHashMap<Long, List<TrackerPayload>> eventsBeingSent = new ConcurrentHashMap<>();
+
+    public InMemoryEventStore() {
+        eventBuffer = new LinkedBlockingDeque<>();
+    }
+
+    public InMemoryEventStore(int bufferCapacity) {
+        eventBuffer = new LinkedBlockingDeque<>(bufferCapacity);
+    }
 
     @Override
-    public boolean add(TrackerPayload trackerPayload) {
+    public boolean addEvent(TrackerPayload trackerPayload) {
         return eventBuffer.offer(trackerPayload);
     }
 
     @Override
-    public List<TrackerPayload> removeEvents(int numberToRemove) {
-        // if numberToRemove is greater than the number of events present,
-        // it will return all the events (there's no error)
-        List<TrackerPayload> eventsList = new ArrayList<>();
-        eventBuffer.drainTo(eventsList, numberToRemove);
-        return eventsList;
+    public BatchPayload getEventBatch(int numberToGet) {
+        List<TrackerPayload> eventsToSend = new ArrayList<>();
+
+        eventBuffer.drainTo(eventsToSend, numberToGet);
+
+        // The batch of events is wrapped as a BatchPayload
+        // They're also added to the "pending" event buffer, the eventsBeingSent HashMap
+        BatchPayload batchedEvents = new BatchPayload(batchId.getAndIncrement(), eventsToSend);
+        eventsBeingSent.put(batchedEvents.getBatchId(), batchedEvents.getPayloads());
+        return batchedEvents;
     }
 
     @Override
-    public int getSize() {
-        return eventBuffer.size();
+    public void cleanupAfterSendingAttempt(boolean successfullySent, long batchId) {
+        // Events that successfully sent are deleted from the pending buffer
+        List<TrackerPayload> events = eventsBeingSent.remove(batchId);
+
+        // Events that didn't send are inserted at the head of the eventBuffer
+        // for immediate resending.
+        if (!successfullySent) {
+            while (events.size() > 0) {
```

Comment:
This bit needs work. If the queue is full, and we want to drop the newer events, then this method must be able to hold the eventBuffer long enough to removeLast and offerFirst all the failed events. Otherwise more events will arrive and it will get full again.

Reply:
yeah, we can add a GH issue. This is not a regression as the same problem was present before. In theory we could check the size before to add considering not only the queue size but also the pending events. It seems a bit too much for this PR so I would just move this problem in a separate issue.

Reply:
Issue #308